---
title: "Лабораторна робота №12. Об'єктно-орієнтоване програмування. Частина ІІ"
description:
  Документ зроблено в [Quarto](https://quarto.org/)
author: "Сідєльніков Даніїл, 2023"
date: "05.23.2023"
lang: ukr
format:
  html:
    code-fold: true
    toc: true # меню
    toc_float: # спливаюче меню  
      collapsed: true # fdnj
      number_sections: true
jupyter: python3
---

__Мета:__ _навчитися програмувати в стилі об'єектно-орієнтованого програмування._

::: callout-note
## Примітка
У лабораторній роботі використано матеріали [python institute.](https://pythoninstitute.org/pcep)
:::

#### Що ви будете вміти?
* створювати класи та об'єкти у Python;
* реалізувати механізми інкапсуляції та наслідування;
* визначати та використовувати змінні екземпляра;
* визначати та використовувати методи;
* використовувати композицію.

## Основи об'єктно-орієнтованого програмування на Python. Метод __str()__

::: callout-important
## Важливо

Перед виконанням лабораторної роботи потрібно опрауювати матеріал лекції 13.
:::

## Використання композиції

### Завдання 1

Тепер ми збираємося вбудувати клас `Point` (див. попереднє завдання) в інший клас. Крім того, ми збираємося помістити три точки в один клас, що дозволить нам визначити трикутник. Як ми можемо це зробити?

Новий клас буде називатися `Triangle`, і ось, що нам потрібно:

* конструктор приймає три аргументи -- вони є об'єктами класу `Point`;
* точки зберігаються всередині об'єкта як список;
* клас надає метод без параметрів з назвою `perimeter()`, який обчислює периметр трикутника, описаного трьома точками; периметр – це сума довжин усіх сторін (це просто нагадування, хоча ми впевнені, що Ви самі це чудово знаєте).

Доповніть шаблон, який ми надали в редакторі. Запустіть свій код і перевірте, чи збігається оцінюваний периметр з нашим.

Нижче Ви можете скопіювати код класу `Point`, який ми використовували у попередній лабораторній роботі:

In [1]:
import math

class Point:
    def __init__(self, x=0.0, y=0.0):
        self.__x = x
        self.__y = y

    def get_x(self):
        return self.__x
    
    def get_y(self):
        return self.__y

    def distance_from_xy(self, x, y):
        return math.hypot(self.__x - x, self.__y - y)

    def distance_from_point(self, point):
        return self.distance_from_xy(point.get_x(), point.get_y())
    
class Triangle:
    def __init__(self, point_a, point_b, point_c):
        self.__points = [point_a, point_b, point_c]
    
    def perimeter(self):
        sum = 0
        for i in range(3):
            sum += self.__points[i].distance_from_point(self.__points[(i+1)%3])
        return sum
    
triangle = Triangle(Point(0, 0), Point(1, 0), Point(0, 1))
print(triangle.perimeter())

3.414213562373095


Очікуваний вивод:  

`3.414213562373095`


Цей код включає клас `Poin`t з попереднього завдання, а також додає клас `Triangle`, який містить конструктор для приймання трьох вершин трикутника і метод `perimeter`, що обчислює периметр трикутника, використовуючи відстані між вершинами.

### Контрольні запитання

Припустимо, що наступний фрагмент коду успішно виконано:

In [3]:
class Dog: 
    kennel = 0 
    def __init__(self, breed): 
        self.breed = breed 
        Dog.kennel += 1 

    def __str__(self): 
        return self.breed + " says: Woof!" 


class SheepDog(Dog): 
    def __str__(self): 
        return super().__str__() + " Don't run away, Little Lamb!" 


class GuardDog(Dog): 
     def __str__(self): 
         return super().__str__() + " Stay where you are, Mister Intruder!" 

rocky = SheepDog("Collie") 
luna = GuardDog("Dobermann")

1. Який очікуваний вивід наступного фрагмента коду?

`print(rocky)` 
`print(luna)`

2. Який очікуваний вивід наступного фрагмента коду?

`print(issubclass(SheepDog, Dog), issubclass(SheepDog, GuardDog))`   
`print(isinstance(rocky, GuardDog), isinstance(luna, GuardDog))`

1. Collie says: Woof! Don't run away, Little Lamb!<br>
Dobermann says: Woof! Stay where you are, Mister Intruder!

2. True False<br>
   False True

### References

1. [Anaconda (Python distribution)](https://uk.wikipedia.org/wiki/Anaconda_(Python_distribution))  
1. [Conda](https://conda.io/en/latest/)  
1. [Pro Git Book](https://git-scm.com/book/en/v2)  
1. [OpenEDG Python Institute](https://pythoninstitute.org/)  
1. [Cisco. Networking Academy](https://www.netacad.com/)  
1. [Научно-издательская система Quarto](https://data-visualization-blog.netlify.app/posts/quarto/)
1. [The Python Standard Library](https://docs.python.org/3/library/index.html)
1. [Callout Blocks. Markdown Syntax](https://quarto.org/docs/authoring/callouts.html)  
1. [5. Built-in Exceptions](https://docs.python.org/3.6/library/exceptions.html)
1. [Принцип єдиної відповідальності](https://uk.wikipedia.org/wiki/%D0%9F%D1%80%D0%B8%D0%BD%D1%86%D0%B8%D0%BF_%D1%94%D0%B4%D0%B8%D0%BD%D0%BE%D1%97_%D0%B2%D1%96%D0%B4%D0%BF%D0%BE%D0%B2%D1%96%D0%B4%D0%B0%D0%BB%D1%8C%D0%BD%D0%BE%D1%81%D1%82%D1%96)
